docs/section-7/swerve-drive.md

```diff
@@ -0,0 +1,164 @@
+
+<a title="Truthdowser at English Wikipedia / CC BY (https://creativecommons.org/licenses/by/3.0)" href="https://commons.wikimedia.org/wiki/File:Wheelbase_and_Track.png"><img width="512" alt="Wheelbase and Track" src="https://upload.wikimedia.org/wikipedia/commons/5/52/Wheelbase_and_Track.png"></a>
+
+Next, we want to calculate the [`SwerveModuleState`](https://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/kinematics/SwerveModuleState.html) for each corner of the drivetrain. The way we do this depends on if our speeds are robot relative or field relative. If they are field relative, then we need to use [`ChassisSpeeds.fromFieldRelativeSpeeds`](https://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/kinematics/ChassisSpeeds.html#fromFieldRelativeSpeeds(double,double,double,edu.wpi.first.wpilibj.geometry.Rotation2d)) and pass in our current heading.
+
+```
+ChassisSpeeds speeds;
+if (SmartDashboard.getBoolean("Field Relative", true)) {
+    speeds = ChassisSpeeds.fromFieldRelativeSpeeds(forward, strafe, rotation, 
+                                                   Rotation2d.fromDegrees(getHeading()));
+} else {
+    speeds = new ChassisSpeeds(forward, strafe, rotation);
+}
+SwerveModuleStates states[] = kinematics.toSwerveModuleStates(speeds);
+```
+
+Where forward, strafe, and rotation are our desired forward/backward speed in m/s, left/right speed in m/s, and counterclockwise angular speed in rad/s respectively.
+
+What is a `SwerveModuleState`? It is an object that stores the speed, in m/s, at which to drive at and the angle, as a `Rotation2d` object, at which to orient the swerve module. So, `SwerveModuleState state = new SwerveModuleState(3.0, Rotation2d.fromDegrees(45));` represents a module running at 3.0 m/s facing 45 degrees. Simply put, a `SwerveModuleState` object represents a velocity vector for a particular swerve module.
+
+Finally, we use our `SwerveModuleState` array to drive our motors:
+
+```
+/* Ensure that the speeds in the array of states are less than the maxSpeed of the robot, 
+   but also ensure the ratio between speeds is the same. */
+SwerveDriveKinematics.normalizeWheelSpeeds(states, maxSpeed);
+moduleFL.move(states[0].speedMetersPerSecond / maxSpeed, states[0].angle.getRadians() / (2 * Math.PI));
+moduleFR.move(states[1].speedMetersPerSecond / maxSpeed, states[1].angle.getRadians() / (2 * Math.PI));
+/* Do the same for the rest of the modules. It is recommended that you create a 
+SwerveModule class which handles moving the motors for a particular swerve module. */"
```

Comment:
Recommend making it clear the `moduleFL` and `moduleFR` are instances of the `SwerveModule` class that the reader would need to write. Perhaps *start* with something, "All that's left is to take info in the `SwerveModuleState` objects and use it to control the swerve modules. It's recommended that you create a separate class, perhaps named `SwerveModule`, for that purpose. Below, `moduleFL` and `moduleFR` are instances of such a class."